contrib/terraform/upcloud/README.md

```diff
@@ -108,6 +81,8 @@ terraform destroy --var-file cluster-settings.tfvars \
   * `cpu`: number of cpu cores
   * `mem`: memory size in MB
   * `disk_size`: The size of the storage in GB
+  * `force_public_ip`: If `use_public_ips` is set to `false`, this forces a public NIC onto the machine anyway when set to `true`. Useful if you're migrating from public nodes to only private. Defaults to `false`
+  * `force_no_user_data`: If `private_network_dns` is set, existing nodes will be recreated since it will try to update their user_data. This forces this machine to not add the user_data and thus will not be recreated. Useful if you're migrating from public nodes to only private. Defaults to `false`
```

Comment:
`force_public_ip` looks to be useful even when not migrating (for example, if you want to have a subset of your machines on the public network). However, `force_no_user_data` looks like it's only introduced to prevent existing resources from being deleted.

To me it looks like `private_network_dns` is actually letting you configure any DNS server that you want. It doesn't necessarily have to be for private network purposes? If that is the case, what about:
* global config option `dns_servers`.
* machine level config option `dns_servers`.

Then you could override `dns_servers` on the machine level and set it to empty list and conditionally render user_data only if `dns_servers` is not an empty list.

Reply:
Yea, I like that approach. I think I was too much in my head with the migration that I didn't think about that.

The only thing that I'm thinking is that this approach is also only there because the DNS via DHCP isn't working on private networks. So when/if they add that feature I'd like the global `dns_servers` to be that. But then again, if we have the per-node override then we can solve existing machines with that one anyway.

Reply:
@simonklb Is this what you expected https://github.com/elastisys/kubespray/pull/20/commits/b17dc2029df73495375dfb6bd7af703b493d6e91

Reply:
Yes! But now you will create user_data in both cases though, right?

Reply:
Sorry, nevermind. It's rendered before it's set in the machine config. Nice :+1:

Reply:
If you haven't checked it it might be worth doing so. I'm not sure if an empty list will still introduce newlines and not equal to `""`. If that is the case it would be easy to just use some trim whitespace function or change the logic to only render if length > 0.

Reply:
I've tested and it will render as `""` so it should be fine.